### PR TITLE
UCP/WIREUP: prefer AUX transport with async active message callback

### PR DESCRIPTION
**Pull request by changchengx**
## What
1. prefer selecting AUX transport with async active message callback
2. add gtest case to check select TCP as AUX transport.

## Why ?
The current criteria requires the tl has UCT_IFACE_FLAG_CB_ASYNC capability(UCP_ADDR_IFACE_FLAG_CB_ASYNC), it results that TCP can't be selected as AUX transport and rc_verbs can't exchange qp info when it's limited to use TCP and rc_verbs transport in connect-to-worker model.
This PR enables selecting the tls to be used as AUX transport if there's no tls has UCT_IFACE_FLAG_CB_ASYNC capability.

### CODE REVIEW

**Comment by changchengx** [comment]
For the failed test case **UCX PR(io_demo Test active messages on CX6/RC)**, the [log](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/62775/logs/689) shows it's not related with this PR
```
2023-05-09T10:58:43.4024034Z         extra_env_server : ''
2023-05-09T10:58:43.7054541Z swx-rain02: bash: /scrap/azure/agent-01/AZP_WORKSPACE/1/s/iodemo_commands_swx-rain02.sh: No such file or directory
2023-05-09T10:58:43.7062699Z pdsh@swx-rain01: swx-rain02: ssh exited with exit code 127
```

**Comment by changchengx** [comment]
For the failed test case **UCX PR(io_demo Test tag match on CX4)**, the [failure log](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/62775/logs/391) shows it's not related with this PR:
```
2023-05-09T10:27:50.1081723Z swx-rain02: bash: /scrap/azure/agent-01/AZP_WORKSPACE/1/s/iodemo_commands_swx-rain02.sh: No such file or directory
2023-05-09T10:27:50.1086722Z pdsh@swx-rain01: swx-rain02: ssh exited with exit code 127
```

**Comment by changchengx** [comment]
For the failed test case **UCX PR(io_demo Test tag match on CX6/DC)**, the [failure log](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/62775/logs/746) shows it's not related with this PR.
```
2023-05-09T11:15:04.2664713Z swx-rain02: bash: /scrap/azure/agent-01/AZP_WORKSPACE/1/s/iodemo_commands_swx-rain02.sh: No such file or directory
2023-05-09T11:15:04.2683963Z pdsh@swx-rain01: swx-rain02: ssh exited with exit code 127
```

**Comment by changchengx** [comment]
For the failed test case **UCX PR(io_demo Test tag match on CX6/RC)**, the [failure log](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/62775/logs/548) shows that it's not related with this RP
```
2023-05-09T10:43:14.0020895Z swx-rain02: bash: /scrap/azure/agent-01/AZP_WORKSPACE/1/s/iodemo_commands_swx-rain02.sh: No such file or directory
2023-05-09T10:43:14.0040688Z pdsh@swx-rain01: swx-rain02: ssh exited with exit code 127
```

**Comment by Alexey-Rivkin** [comment]
All tests passed, but status not reported to Github:
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=62775

**Comment by changchengx** [comment]
@evgeny-leksikov This PR is to enable use TCP as auxiliary transport. All necessary CI test cases have been passed.

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:1010]: pass mandatory_flags as parameter, instead of prefer_cb_async_cap flag

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:2376]: 1. space line before
2. add comment before /* Select auxiliary transport that supports async active message callback */

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:2385]: remove this variable

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:2386]: Add a comment before /* Fallback to an auxiliary transport without async active message callback requirement */

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c]: Done. The ```mandatory_flags``` is either ```1```(select AUX transport with async active message callback) or ```0```

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c]: Done

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c]: Done

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c]: Done

**Comment by changchengx** [comment]
I did force push here because the changed LOC is under 50 lines.

**Comment by brminich** [comment]
@yosefe, @changchengx, can you please wait merging the PR, until wire-compat tests are merged?

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:2386]: align text to fill the line up to 80 chars

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:1018]: but now we add UCP_ADDR_IFACE_FLAG_CB_ASYNC even when has_cm (line 1022) is true

File: src/ucp/wireup/select.c
Comment (yosefe) [inline src/ucp/wireup/select.c:1884]: we need CB_ASYNC here

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c:1018]: Should it let ```ucp_wireup_fill_aux_criteria``` check whether it's CM lane to add the mandatory_flags or not?

In this PR, it assumes that the caller decides whether it's needs to add extra ```mandatory_flags```.
Until now, we still keep the existing logic:
1. [ucp_wireup_select_wireup_msg_lane](https://github.com/openucx/ucx/blob/v1.15.0-rc1/src/ucp/wireup/select.c#L1883)
   The ```ep_init_flags``` indicate it could be CM lane, so this PR set ```mandatory_flags=0``` here.
2. In [ucp_wireup_select_aux_transport](https://github.com/openucx/ucx/pull/9061/files#diff-ff110fb0f2e163038a3eb15545d3b81d9d32c121b79e7a764b72fbb1fa643470R2372-R2389), it set ```mandatory_flags = UCP_ADDR_IFACE_FLAG_CB_ASYNC```, then do fallback if there's such existing tl.

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c:1884]: @yosefe Should it add ```CB_ASYNC``` here?
For the ```wireup msg lane```, the ```ep_init_flags``` have the same flags as CM lane. In the original code logic, it doesn't require the CB_ASYNC capability.
If it adds ```CB_ASYNC``` here, the function [ucp_wireup_select_wireup_msg_lane](https://github.com/openucx/ucx/blob/v1.15.0-rc1/src/ucp/wireup/select.c#L1883) may have different execution code path compared with the original code logic.

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c:1018]: Got it. I'll change the code.

File: src/ucp/wireup/select.c
Comment (changchengx) [reply src/ucp/wireup/select.c:1884]: Please ignore the above comment after reading below comment.
https://github.com/openucx/ucx/pull/9061#discussion_r1193545512

Comment (yosefe) [reply src/ucp/wireup/select.c:1884]: Yes, in order to keep the original logic

**Review by yosefe — COMMENTED** [review]
LGTM
Waiting for wire compat CI

**Comment by yosefe** [comment]
@changchengx is there a CI test that covers the scenario of selecting TCP as an aux transport?

**Comment by changchengx** [comment]
> @changchengx is there a CI test that covers the scenario of selecting TCP as an aux transport?

@yosefe I haven't found it. Should we add one for it? For example, limit ucp_hello_world to use TCP & rc_verbs transport.

**Comment by changchengx** [comment]
@yosefe 
I've added the gtest case
```C
/* Test use tcp as AUX transport */
UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wireup_2sided,
                              tcp_aux, "tcp,rc_verbs")
```
The gtest case could cover the test case.
Does it also need to add new test case with ```ucp_hello_world``` by using ```UCX_TLS=tcp,rc_verbs```?

**Review by yosefe — APPROVED** [review]

**Comment by changchengx** [comment]
@yosefe I planned to squash and rebase this PR after merging the ```wire-compat tests```.
Reason:
1) Now, this PR has two small commits. the first commit depends on the second commit. This is not good. Since the two commits are very small, I'll squash them together.
2) After merging ```wire-compat tests```, I'll rebase on that PR. Keep linearly merging or merging on the newer master branch is friendly to ```git-bisect```.

**Comment by changchengx** [comment]
> LGTM Waiting for wire compat CI
>
@yosefe I'll squash and rebase this PR since https://github.com/openucx/ucx/pull/9273 has been merged. If there's some CI errors, I'll solve them based on the squashed PR.

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@brminich can you pls take a look as well?

File: src/ucp/wireup/select.c
Comment (brminich) [inline src/ucp/wireup/select.c:1015]: maybe use `uint64_t` to be consistent with `criteria->remote_iface_flags.mandatory` type

Comment (changchengx) [reply src/ucp/wireup/select.c]: Done. Force pushed changes.

**Review by brminich — APPROVED** [review]

**Comment by changchengx** [comment]
Below failure in [UCX PR(Tests BlueField on worker2)](https://github.com/openucx/ucx/pull/9061/checks?check_run_id=16141147876) isn't related with this PR. It needs to rerun it again.
```
2023-08-23T12:38:16.4527772Z + taskset -c 4 /scrap/azure/agent-03/AZP_WORKSPACE/1/s/install/bin/ucx_perftest -b /scrap/azure/agent-03/AZP_WORKSPACE/1/s/install/share/ucx/perftest/test_types_short_uct -b /scrap/azure/agent-03/AZP_WORKSPACE/1/s/install/share/ucx/perftest/msg_pow2_short -w 1 -d mlx5_3:1 -b /scrap/azure/agent-03/AZP_WORKSPACE/1/s/install/share/ucx/perftest/transports -p 11500
2023-08-23T12:38:16.5077895Z [1692794296.506879] [swx-rain04-bf1:739398:0]        perftest.c:429  UCX  ERROR server failed. bind() failed: Address already in use
2023-08-23T12:38:16.5083714Z [1692794296.506967] [swx-rain04-bf1:739398:0]        perftest.c:889  UCX  ERROR failed to setup RTE transport: Input/output error
2023-08-23T12:38:31.4624774Z + '[' 0 -eq 1 ']'
2023-08-23T12:38:31.4627013Z + client_pid=739632
2023-08-23T12:38:31.4628717Z + wait 739632
```

**Comment by changchengx** [comment]
Just for record:
1. Passed local test to use TCP as AUX transport which is required by vRDMA project.
2. Code info: ucx/branch/master/head: 37f9323134e0
3. configure ```$ ./contrib/configure-release --prefix=$PWD/install --with-ud=no --with-devx=no --with-rdmacm=no --with-mlx5-dv=no```
4. Test environment:
```
    Host A with LAG/bond(cx6dx with eth2 & eth3, bond ip: 192.168.20.62) and eth0(ip: 192.168.121.68)
    Host B with LAG/bond(cx6dx with eth2 & eth3, bond ip: 192.168.20.61) and eth0(ip: 192.168.121.140)
```
6. Run below command on Host A
```$ UCX_SOCKADDR_TLS_PRIORITY=tcp UCX_TLS=rc_verbs,tcp UCX_NET_DEVICES=mlx5_bond_0:1,eth0 ./install/bin/io_demo```
7. Run below command on Host B
```$ UCX_SOCKADDR_TLS_PRIORITY=tcp UCX_TLS=rc_verbs,tcp UCX_NET_DEVICES=mlx5_bond_0:1,eth0 ./install/bin/io_demo -i 1000 192.168.121.68```

**Pull request merged**
